### PR TITLE
Workaround to specify additional AUTH mechanisms with dash

### LINKED ISSUE #224: Current AUTH hook implementation cannot support mechanisms with dash "-" in the name
The current implementation simply searches for `auth_*` methods in the handler, chop off the `auth_` part, and present the chopped name as is to the list of supported authentications.

This means that AUTH mechanisms such as "CRAM-MD5" and "DIGEST-MD5" are not representable.

Fixing this is a bit tricky; the simple trick of replacing "_" with "-" will result in SASL mechanisms that _does_ contain underscore in their names (see [SASL mechanisms registry on IANA](https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml)) to be not representable.

My suggestion would be to add another decorator (similar to the existing `@syntax` decorator) that will add an attribute, say `__auth_mechanism_name__` ; if the attribute exists, use that. If the attribute does not exist, use the existing name-mangling strategy.

### PR DESCRIPTION
**Pull request by pepoluan**
## What do these changes do?

The current implementation for AUTH Hooks does not allow AUTH mechanisms with dash(es) in their names.

For example, it is not possible to create a handler method for `CRAM-MD5` or `DIGEST-MD5`, because the code simply takes the method name beginning with "auth_" and chop off that part.

This PR provides two methods of specifying dashes:

  * Using "double underscore", so `MECH__NAME__WITH__DASH` ==> `MECH-NAME-WITH-DASH`
  *  Using `@auth_mechanism(actual_name)` decorator

## Are there changes in behavior for the user?

None for existing users unless they had subclassed `SMTP` to implement this feature.

## Related issue number

Closes #224 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] tox testenvs have been executed in the following environments:
  - [x] Windows 10 (via PyCharm runner): `(ALL)`
  - [x] Windows 10 (via PSCore 7.0.3): `(ALL)`
  - [x] Windows 10 (via Cygwin): `qa,py36-{nocov,cov}`
  - [x] Ubuntu 18.04 on WSL 1.0: `(ALL)` + `pypy3-{nocov,cov}`
  - [x] Ubuntu 18.04 on VBox: `(ALL)` + `pypy3-{nocov,cov}`
  - [x] Ubuntu 20.04 on VBox: `(ALL)` + `pypy3-{nocov,cov}`
  - [x] FreeBSD 12.1 on VBox: `(ALL)` + `pypy3-nocov`
- [x] Documentation reflects the changes
- [x] Add a news fragment into the `NEWS.rst` file

### CODE REVIEW

**Comment by pepoluan** [comment]
Will undraft after I do testing on my test systems.

File: aiosmtpd/smtp.py [hunk 1]
```diff
@@ -61,6 +65,7 @@ class _DataState(enum.Enum):
 EMPTYBYTES = b''
 MISSING = _Missing()
 NEWLINE = '\n'
+VALID_AUTHMECH = re.compile(r"[A-Z0-9_-]+\Z")
```
Comment (pepoluan) [inline aiosmtpd/smtp.py:68]: We can revisit this as a bugfix if there is actually a SASL mechanism which has characters outside of these.

**Review by pepoluan — COMMENTED** [review]
Left some comments.

File: .gitignore [hunk 2]
```diff
@@ -30,3 +30,4 @@ _dynamic/
 coverage.xml
 diffcov.html
 .python-version
+.pytype/
```
Comment (pepoluan) [inline .gitignore:33]: Workdir used by `pytype`

File: aiosmtpd/smtp.py [hunk 3]
```diff
@@ -179,21 +224,25 @@ def __init__(
...
         self._auth_required = auth_required
-        self.authenticated = False
         # Get hooks & methods to significantly speedup getattr's
         self._auth_methods: Dict[str, _AuthMechAttr] = {
-            m.replace("auth_", ""): _AuthMechAttr(getattr(h, m), h is self)
-            for h in (self, handler)
-            for m in dir(h)
-            if m.startswith("auth_")
+            getattr(
+                mfunc, "__auth_mechanism_name__",
```
Comment (pepoluan) [inline aiosmtpd/smtp.py:230]: Option 1: Decorated auth hooks

File: aiosmtpd/smtp.py [hunk 4]
```diff
@@ -179,21 +224,25 @@ def __init__(
...
+                mname.replace("auth_", "").replace("__", "-")
```
Comment (pepoluan) [inline aiosmtpd/smtp.py:231]: Option 2: Double-underscored auth hooks

**Review by waynew — APPROVED** [review]
🚀

**Comment by pepoluan** [comment]
I'm gonna let this marinate for 2x24 before squamerging, just in case 😊

**Comment by pepoluan** [comment]
First squamerge of 2021, yay! 😁

**Pull request merged**